### PR TITLE
Introducing KPIEvent and supporting Avro for nakadi.subscription.log

### PR DESCRIPTION
**Pull request by thomasabraham**
# Introducing KPIEvent and supporting Avro for nakadi.subscription.log

> Zalando ticket : team-aruha/issues/984

## Description
A new class called `KPIEvent` is introduced. All KPI events are supposed
to be exted from this. Currently the only KPI event that is published
using this is the `SubscriptionLogEvent`. The `KPIEvent` supports
mapping the classes that extends it directly to `GenericRecord` or
`JSONObject`; as long as the fields are annotated with `@KPIField`
annotation. The `NakadiKpiPublisher` class is also supporting the
publishing of KPI Events now.

## Review
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    /*
+        Implementation of this method should return the event-type name of the specific KPI event.
+        This method should not be confused with the getter "getEventType()" in some specific KPI events;
+        which is why this method is named "eventTypeOfThisKPIEvent()
+     */
+    public abstract String eventTypeOfThisKPIEvent();
+
+    private Stream<Field> getKPIFields() {
+        return Arrays.stream(this.getClass().getDeclaredFields())
```
Comment (adyach) [inline core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java:36]: that works only with a higher level classes, what if I have more sophisticated structures ?

Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java:36]: Yes, the current implementation is only taking care of the simple structure with one level. 
I'm hoping to take care of the more complex structure once I do have a example for it.

File: core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                .filter(field -> field.isAnnotationPresent(KPIField.class));
+    }
+
+    public JSONObject asJsonObject() throws KPIEventMappingException {
+        final var jsonObject = new JSONObject();
+        getKPIFields().forEach(field -> {
+            try {
+                jsonObject.put(field.getAnnotation(KPIField.class).value(), field.get(this));
```
Comment (adyach) [inline core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java:44]: Nice approach with annotations! looks very clean in this way. 

Can optimize it even further - cache fields annotations for each class (usually it is done by creating/initializing mappers once per startup) and avoid using clazz.getFields() for each publishing call (there are thousands per second publishes).

Another thing - maybe we can avoid injecting mappers logic into objects (looks like mixing it alltogether, plus inheritance actually removes the beauty of annotations usage, maybe it is also possible to improve).

Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java:44]: Good point, I'll do this.

**Comment by thomasabraham** [comment]
:+1:

File: core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java (hunk 2 above)
Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java]: Done

File: core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                .map(field -> findGetterByField(mappedClass, field))
+                .collect(Collectors.toSet());
+    }
+
+    private KPIGetter findGetterByField(final Class<? extends KPIEvent> clazz, final Field field) {
+        final var kpiField = field.getAnnotation(KPIEvent.KPIField.class);
+        final String getterName;
+        if (kpiField.getter().isEmpty()) {
+            final var fieldName = field.getName();
+            getterName = "get" + fieldName.substring(0, 1).toUpperCase() + fieldName.substring(1);
```
Comment (adyach) [inline core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java:61]: this is quite dangerous. what about `isRunning()`, upper case... can you just get all the methods. 
for example, Jackson in some places makes [such checks](https://github.com/FasterXML/jackson-databind/blob/39885e702795dc4d57f416b3e44ea1f86d33d5af/src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertiesCollector.java#L680) to get differentiate getters / setters 

but ultimately it is better to use existing utils to access fields, getters etc. e.g. BeanUtils from Spring or Apache commons. however, I have limited experience writing such logic.

File: core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                throw new KPIEventMappingException("Cannot map KPIEvent " + kpiEvent.getClass().getName()
+                        + " to JSON. Unable to read " + kpiGetter.name, iae);
+            }
+        });
+        return jsonObject;
+    }
+
+    private Set<KPIGetter> gettersOf(final Class<? extends KPIEvent> clazz) {
+        if (!this.fieldGetterCache.containsKey(clazz)) {
+            synchronized (this) {
```
Comment (adyach) [inline core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java:101]: I am not sure why do you have it? we should make sure that we have loaded all possible kpi events before we start, then there will be no need to cache on the fly

File: core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiKpiPublisher.java [hunk 5]
```diff
@@ -49,6 +53,42 @@ protected NakadiKpiPublisher(
         this.eventMetadata = eventMetadata;
         this.uuidGenerator = uuidGenerator;
         this.avroSchema = avroSchema;
+        this.kpiEventMapper = new KPIEventMapper(Set.of(SubscriptionLogEvent.class));
+    }
+
+    public void publish(final Supplier<KPIEvent> kpiEventSupplier) {
```
Comment (adyach) [inline core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiKpiPublisher.java:59]: I miss usage of the method

File: core-common/src/main/java/org/zalando/nakadi/domain/kpi/SubscriptionLogEvent.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package org.zalando.nakadi.domain.kpi;
+
+import org.zalando.nakadi.config.KPIEventTypes;
+
+public class SubscriptionLogEvent extends KPIEvent {
```
[the remark this answers is not in the document]
Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/domain/kpi/SubscriptionLogEvent.java:5]: Extending allows it to be passed to the `publish(final Supplier<KPIEvent> kpiEventSupplier)` method.
While annotation is used to indicate the fields.

[the remark this answers is not in the document]
Comment (adyach) [reply core-common/src/main/java/org/zalando/nakadi/domain/kpi/SubscriptionLogEvent.java:5]: missed that, all good then

File: core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java (hunk 3 above)
Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java:61]: I'll continue to improve this. I didn't want to over engineer with all the cases in the first version itself; especially without an example.

File: core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java (hunk 4 above)
Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java:101]: We plan to load it all during start, but this is there to make the code still work if we forgot to add one.

Comment (adyach) [reply core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java:101]: can we avoid that logic for now and fail if the code meets unrecognised bean ?

File: core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiKpiPublisher.java (hunk 5 above)
Comment (thomasabraham) [reply core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiKpiPublisher.java:59]: Good point; forgot to use the new feature.

**Comment by thomasabraham** [comment]
:+1:

File: core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java (hunk 4 above)
Comment (thomasabraham) [reply core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java]: Done

File: core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiKpiPublisher.java (hunk 5 above)
Comment (thomasabraham) [reply core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiKpiPublisher.java:59]: This is done now.

**Comment by adyach** [comment]
👍

**Comment by thomasabraham** [comment]
:+1:

**Pull request merged**